### PR TITLE
composite checkout - add plan length picker to review step

### PR DESCRIPTION
**Pull request by nbloomf**
I recommend looking at the commits one at a time.

#### Changes proposed in this Pull Request

* Adds radio buttons to the order review step of composite checkout allowing users to change the plan term length.

To do:

- [x] Use a state machine to enable immediate feedback

#### Testing instructions

```
npm run test-client client/my-sites/checkout/checkout
```

Manual testing:
1. Enter checkout with a plan
2. Proceed to the review order step
3. Verify that radio buttons appear corresponding to the plan term lengths (1 year and 2 years), and that the currently selected term is correct.
4. Click to select a different term length; verify that the selected radio button and the prices (line item and total) update.

Edge cases:
1. Change the plan length, then change it again before the first change takes effect.
2. Switch between plan lengths several times in succession

Example:

image

image

### CODE REVIEW

File: client/my-sites/checkout/checkout/composite-checkout.js [hunk 1]
```diff
@@ -155,6 +162,89 @@ export default function CompositeCheckout( {
 		]
 	);
 
+	function getTermText( term ) {
+		switch ( term ) {
+			case TERM_BIENNIALLY:
+				return translate( 'Two years', {
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/composite-checkout.js]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **8** times:
`translate( 'Two years' )` *ES Score: 14*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/plan-length-picker-to-composite-checkout)


:information_source: This string already exists without a **context**. Only add a context if the meaning of the string is very specific.

File: client/my-sites/checkout/checkout/composite-checkout.js [hunk 2]
```diff
@@ -155,6 +162,89 @@ export default function CompositeCheckout( {
...
+					context: 'subscription length',
+				} );
+
+			case TERM_ANNUALLY:
+				return translate( 'One year', {
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/composite-checkout.js]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **9** times:
`translate( 'One year' )` *ES Score: 12*


:information_source: This string already exists without a **context**. Only add a context if the meaning of the string is very specific.

File: client/my-sites/checkout/checkout/composite-checkout.js [hunk 3]
```diff
@@ -155,6 +162,89 @@ export default function CompositeCheckout( {
...
+					context: 'subscription length',
+				} );
+
+			case TERM_MONTHLY:
+				return translate( 'One month', {
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/composite-checkout.js]: :information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

File: client/my-sites/checkout/checkout/composite-checkout.js [hunk 4]
```diff
@@ -194,6 +201,119 @@ export default function CompositeCheckout( {
 		);
 	};
 
+    function getTermText( term ) {
+        switch ( term ) {
+            case TERM_BIENNIALLY:
+                return translate( 'Two years', {
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/composite-checkout.js]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **8** times:
`translate( 'Two years' )` *ES Score: 14*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/plan-length-picker-to-composite-checkout)


:information_source: This string already exists without a **context**. Only add a context if the meaning of the string is very specific.

File: client/my-sites/checkout/checkout/composite-checkout.js [hunk 5]
```diff
@@ -194,6 +201,119 @@ export default function CompositeCheckout( {
...
+                    context: 'subscription length',
+                } );
+
+            case TERM_ANNUALLY:
+                return translate( 'One year', {
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/composite-checkout.js]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **9** times:
`translate( 'One year' )` *ES Score: 12*


:information_source: This string already exists without a **context**. Only add a context if the meaning of the string is very specific.

File: client/my-sites/checkout/checkout/composite-checkout.js [hunk 6]
```diff
@@ -194,6 +201,119 @@ export default function CompositeCheckout( {
...
+                    context: 'subscription length',
+                } );
+
+            case TERM_MONTHLY:
+                return translate( 'One month', {
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/composite-checkout.js]: :information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

File: client/my-sites/checkout/checkout/composite-checkout.js [hunk 7]
```diff
@@ -390,3 +403,143 @@ function TestingBanner() {
...
+			return savings > 0 ? <Discount>-{ savings.toString() }%</Discount> : null;
+		};
+
+		return productsWithPrices.map( variant => {
+			const label = getTermText( variant.plan.term, translate );
+			const price = (
+				<React.Fragment>
+					{ percentSavings( variant.priceMonthly ) }
+					{ variant.product.cost_display }
+					{ getTaxText( translate ) }
```
Comment (fditrapani) [inline client/my-sites/checkout/checkout/composite-checkout.js]: Since we're showing taxes in the summary, this doesn't seem necessary to me.

**Comment by sirbrillig** [comment]
Works as described. I'll withhold my final approval until those linting errors are fixed... actually! `dispatch` is totally safe to add (you should still do it), but `productSlug` is only there because of a `debug()` call, which you could remove. I'd recommend that.

**Comment by nbloomf** [comment]
Thanks for testing @fditrapani! Regarding

> discrepancy in the pricing

That is strange; I'm not able to reproduce this after changing my currency to CAD. We definitely need to figure out what's happening though. It would be helpful to have the network request and response for the call to `me/shopping-cart` when the wrong pricing is displayed. FWIW, it looks like the term option prices are correct but the item and total are not. 276 is not the full price of any plan in any currency.

**Comment by nbloomf** [comment]
I went ahead and added the crossed out "two years at one year price" label for two year plans.

image

**Comment by fditrapani** [comment]
> I went ahead and added the crossed out "two years at one year price" label for two year plans.

Nice! That looks much clearer. Thanks!

> It would be helpful to have the network request and response for the call to me/shopping-cart when the wrong pricing is displayed.

I tried taking a look but I couldn't find anything. What should I be looking for in the network tab?

**Comment by nbloomf** [comment]
After writing that response I realized this was probably a case of proration not being included in the variant prices. That's a much hairier problem to solve, so for now I've just suppressed the option for non-signup purchases in #39439.

**Pull request merged**